pkg/lib/authn/otp/magic_link_store_redis.go

```diff
@@ -0,0 +1,89 @@
+package otp
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	goredis "github.com/go-redis/redis/v8"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/lib/infra/redis/appredis"
+	"github.com/authgear/authgear-server/pkg/util/clock"
+)
+
+type MagicLinkStoreRedis struct {
+	Redis *appredis.Handle
+	AppID config.AppID
+	Clock clock.Clock
+}
+
+func (s *MagicLinkStoreRedis) set(token string, target string, expireAt time.Time) error {
+	ctx := context.Background()
+	data, err := json.Marshal(target)
+	if err != nil {
+		return err
+	}
+
+	return s.Redis.WithConn(func(conn *goredis.Conn) error {
+		key := redisMagicLinkKey(s.AppID, token)
+		ttl := expireAt.Sub(s.Clock.NowUTC())
+
+		_, err := conn.SetEX(ctx, key, data, ttl).Result()
```

Comment:
We should use `SetNX` for this case, we should not allow overwritten this mapping. Although the magic link code is relatively long, we cannot guarantee its uniqueness. If we generate a duplicated code, the original mapping will be overwritten and the verification will go to an unexpected state.

So if it is duplicated, we should return err which is done in the next 2 lines.